### PR TITLE
[eclipse/xtext#1679] Refactor more Xtend to java.

### PR DESCRIPTION
**Pull request by ArneDeutsch**
Signed-off-by: Arne Deutsch <Arne.Deutsch@itemis.de>

### CODE REVIEW

**Comment by cdietrich** [comment]
please check for the test fails

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java [hunk 1]
```diff
@@ -217,52 +141,44 @@ public URI toPortableURI(StorageAwareResource sourceResource, EObject targetObje
...
-		if (desc.descriptionRelativeFragment != null) {
+				+ URI.encodeFragment(join(segments, ":"), false);
+		if (desc.descriptionRelativeFragment != null)
 			uriFragment += "#" + URI.encodeFragment(desc.descriptionRelativeFragment, false);
-		}
 		return uriFragment;
 	}
 
-	protected PortableURIs.PortableFragmentDescription fromFragmentString(String fragmentString) {
+	protected PortableFragmentDescription fromFragmentString(String fragmentString) {
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java:171]: I prefer the version on master.

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java [hunk 2]
```diff
@@ -217,52 +141,44 @@ public URI toPortableURI(StorageAwareResource sourceResource, EObject targetObje
...
-			}
-		}
-		QualifiedName qname = QualifiedName.create(Splitter.on(":").splitToList(URI.decode(segments.next())));
+		EPackage ePackage = packageRegistry.getEPackage(eClassURI.trimFragment().toString());
+		Resource eResource = ePackage == null ? null : ePackage.eResource();
+		EClass eClass = (EClass) (eResource == null ? null : eResource.getEObject(eClassURI.fragment()));
+		QualifiedName qname = QualifiedName.create(toList(Splitter.on(":").split(URI.decode(segments.next()))));
 		String fragment = segments.hasNext() ? URI.decode(segments.next()) : null;
-		return new PortableURIs.PortableFragmentDescription(eClass, qname, fragment);
+		return new PortableFragmentDescription(eClass == null ? EcorePackage.Literals.EOBJECT : eClass, qname,
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java:180]: Could use the short class name for the inner type, though.

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java [hunk 3]
```diff
@@ -276,10 +192,21 @@ protected String toFragmentString(PortableURIs.PortableFragmentDescription desc)
 	 * @return a fragment path from the given container to the child, or <code>null</null> is fromContainer == toChild
 	 * 
 	 * @see #getEObject(EObject,String)
-	 * @throws IllegalArgumentException if the child is not a child of the given container.
 	 */
-	public String getFragment(EObject fromContainer, EObject toChild) throws IllegalArgumentException {
-		return Strings.emptyToNull(EcoreUtil.getRelativeURIFragmentPath(fromContainer, toChild));
+	public String getFragment(EObject fromContainer, EObject toChild) {
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java:196]: This reverts the cleanup that is present on master.

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	/**
+	 * Finds or creates a ResourceStorageLoadable for the given resource. Clients should first call
+	 * shouldLoadFromStorage to check whether there exists a storage version of the given resource.
+	 * 
+	 * @return an IResourceStorageLoadable
+	 */
+	@Override
+	public ResourceStorageLoadable getOrCreateResourceStorageLoadable(StorageAwareResource resource) {
+		try {
+			ResourceStorageProviderAdapter stateProvider = head(
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.java:68]: IterableExtensions are still being used here. Can this be simplified, e.g. by using FluentIterable or streams?

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	@Override
+	public boolean hasStorageFor(URI uri) {
+		return new ExtensibleURIConverterImpl().exists(getBinaryStorageURI(uri), Collections.emptyMap());
+	}
+
+	protected URI getBinaryStorageURI(URI sourceURI) {
+		return sourceURI.trimSegments(1).appendSegment("." + sourceURI.lastSegment() + "bin");
+	}
+
+	@Pure
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.java:152]: `@Pure` can be removed

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+public class ResourceStorageFacade implements IResourceStorageFacade {
+	private static final Logger LOG = Logger.getLogger(ResourceStorageFacade.class);
+
+	@Inject
+	private IContextualOutputConfigurationProvider outputConfigurationProvider;
+
+	@Inject
+	private Provider<AbstractFileSystemAccess2> fileSystemAccessProvider;
+
+	@Accessors
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.java:45]: `@Accessors` can be removed

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+	@Override
+	public ResourceStorageWritable createResourceStorageWritable(OutputStream out) {
+		return new ResourceStorageWritable(out, isStoreNodeModel());
+	}
+
+	/**
+	 * @return whether a stored resource state exists for the given resource
+	 */
+	protected boolean doesStorageExist(StorageAwareResource resource) {
+		ResourceStorageProviderAdapter stateProvider = head(
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.java:113]: EcoreUtil.getAdapter?

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.xtend [hunk 8]
```diff
@@ -1,151 +0,0 @@
...
-	}
-	
-	override void saveResource(StorageAwareResource resource, IFileSystemAccessExtension3 fsa) {
-		val path = computeOutputPath(resource)
-		val bout = new MyByteArrayOutputStream()
-		val outStream = createResourceStorageWritable(bout)
-		try {
-			outStream.writeResource(resource)
-		} catch(IOException e) {
-			// something went wrong when writing the resource - stream's content is bogus and not written to disk
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.xtend:83]: Comment got lost

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.xtend [hunk 9]
```diff
@@ -1,151 +0,0 @@
...
-	}
-	
-	/**
-	 * @return whether a stored resource state exists for the given resource
-	 */
-	protected def doesStorageExist(StorageAwareResource resource) {
-		val stateProvider = resource.resourceSet.eAdapters.filter(ResourceStorageProviderAdapter).head
-		if (stateProvider !== null && stateProvider.getResourceStorageLoadable(resource) !== null)
-			return true;
-		// check for next to original location, i.e. jars
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.xtend:105]: Comments got lost

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageLoadable.xtend [hunk 10]
```diff
@@ -1,113 +0,0 @@
...
-			zipIn.nextEntry
-			readNodeModel(resource, new BufferedInputStream(zipIn))
-		}
-	}
-	
-	protected def void readContents(StorageAwareResource resource, InputStream inputStream) throws IOException {
-		val in = new BinaryResourceImpl.EObjectInputStream(inputStream, emptyMap) {
-			
-			override readCompressedInt() throws IOException {
-				//HACK! null resource set, to avoid usage of resourceSet's package registry
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageLoadable.xtend:71]: Comment got lost

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageLoadable.xtend [hunk 11]
```diff
@@ -1,113 +0,0 @@
...
-	protected def void readResourceDescription(StorageAwareResource resource, InputStream inputStream) throws IOException {
-		val objectIn = new ObjectInputStream(inputStream)
-		val description = objectIn.readObject as SerializableResourceDescription
-		description.updateResourceURI(resource.URI)
-		resource.resourceDescription = description
-	}
-	
-	protected def void readNodeModel(StorageAwareResource resource, InputStream inputStream) throws IOException {
-		val serializableNodeModel = new SerializableNodeModel(resource)
-		// if this is a synthetic resource (i.e. tests or so, don't load the node model)
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageLoadable.xtend:99]: Comment got lost

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageProviderAdapter.java [hunk 12]
```diff
@@ -19,7 +19,7 @@
 public class ResourceStorageProviderAdapter extends AdapterImpl {
 	@Override
 	public boolean isAdapterForType(Object type) {
-		return ResourceStorageProviderAdapter.class.equals(type);
+		return type == ResourceStorageProviderAdapter.class;
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageProviderAdapter.java:22]: I prefer equals

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableResourceDescription.java [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+ */
+public class SerializableResourceDescription extends AbstractResourceDescription implements Externalizable {
+
+	private List<SerializableEObjectDescription> descriptions = Collections.emptyList();
+
+	private List<SerializableReferenceDescription> references = Collections.emptyList();
+
+	private List<QualifiedName> importedNames = Collections.emptyList();
+
+	private URI uRI;
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableResourceDescription.java:45]: fieldname should be `uri`

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableResourceDescription.java [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+
+	public void setImportedNames(List<QualifiedName> importedNames) {
+		this.importedNames = importedNames;
+	}
+
+	public URI getURI() {
+		return uRI;
+	}
+
+	public void setURI(URI uRI) {
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableResourceDescription.java:127]: parametername should be `uri`

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SourceLevelURIsAdapter.java [hunk 15]
```diff
@@ -8,11 +8,12 @@
  */
 package org.eclipse.xtext.resource.persistence;
 
+import static org.eclipse.xtext.xbase.lib.IterableExtensions.*;
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SourceLevelURIsAdapter.java:11]: I prefer the version on master

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java [hunk 16]
```diff
@@ -46,27 +45,25 @@
 	public void load(Map<?, ?> options) throws IOException {
 		if (!isLoaded && !isLoading && resourceStorageFacade != null
 				&& resourceStorageFacade.shouldLoadFromStorage(this)) {
-			LOG.debug("Loading " + getURI() + " from storage.");
+			if (StorageAwareResource.LOG.isDebugEnabled())
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java:48]: I prefer the version on master

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResourceDescriptionManager.java [hunk 17]
```diff
@@ -21,10 +21,9 @@
 	@Override
 	public IResourceDescription getResourceDescription(Resource resource) {
 		if (resource instanceof StorageAwareResource) {
-			IResourceDescription result = ((StorageAwareResource) resource).getResourceDescription();
-			if (result != null) {
-				return result;
-			}
+			IResourceDescription resourceDescription = ((StorageAwareResource) resource).getResourceDescription();
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResourceDescriptionManager.java:24]: I prefer the version on matser

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/IResourceStorageFacade.java [hunk 18]
```diff
@@ -22,37 +22,36 @@
...
+	boolean shouldLoadFromStorage(StorageAwareResource resource);
+
+	/**
+	 * @return whether storage data exists for the given URI
+	 */
+	boolean hasStorageFor(URI uri);
+
+	/**
+	 * Finds or creates a ResourceStorageLoadable for the given resource. Clients should first call
+	 * shouldLoadFromStorage to check whether there exists a storage version of the given resource.
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/IResourceStorageFacade.java:37]: {@link #shouldLoadFromStorage}

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/IResourceStorageFacade.java [hunk 19]
```diff
@@ -22,37 +22,36 @@
...
+	 * 
+	 * @return an IResourceStorageLoadable
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/IResourceStorageFacade.java:39]: Doesn't agree with the actual return type, and it's a useless comment

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/IResourceStorageFacade.java [hunk 20]
```diff
@@ -22,37 +22,36 @@
...
+	 */
+	ResourceStorageLoadable getOrCreateResourceStorageLoadable(StorageAwareResource resource);
+
+	/**
+	 * Saves the resource using the given file system access.
+	 */
+	void saveResource(StorageAwareResource resource, IFileSystemAccessExtension3 fsa);
+
+	/**
+	 * Creates a fresh ResourceStorageWritable wrapping the given OutputStream
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/IResourceStorageFacade.java:49]: @link these types please

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/IResourceStorageFacade.java [hunk 21]
```diff
@@ -22,37 +22,36 @@
...
+	 */
+	ResourceStorageWritable createResourceStorageWritable(OutputStream outputStream);
+
+	/**
+	 * Creates a fresh ResourceStorageLoadable wrapping the given InputStream
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/IResourceStorageFacade.java:54]: @link these types please

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.java [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+		if (adapter == null || adapter.getSourceLevelURIs().contains(resource.getURI()))
+			return false;
+		return doesStorageExist(resource);
+	}
+
+	/**
+	 * Finds or creates a ResourceStorageLoadable for the given resource. Clients should first call
+	 * shouldLoadFromStorage to check whether there exists a storage version of the given resource.
+	 * 
+	 * @return an IResourceStorageLoadable
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.java:63]: Doesn't agree with actual type and is a pretty useless comment

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.java [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+		}
+	}
+
+	@Override
+	public void saveResource(StorageAwareResource resource, IFileSystemAccessExtension3 fsa) {
+		MyByteArrayOutputStream bout = new MyByteArrayOutputStream();
+		try {
+			createResourceStorageWritable(bout).writeResource(resource);
+		} catch (IOException e) {
+			ResourceStorageFacade.LOG.warn("Cannot write storage for " + resource.getURI(), e);
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.java:93]: s/ResourceStorageFacade.LOG/LOG/

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.java [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+	@Override
+	public boolean shouldLoadFromStorage(StorageAwareResource resource) {
+		SourceLevelURIsAdapter adapter = SourceLevelURIsAdapter.findInstalledAdapter(resource.getResourceSet());
+		if (adapter == null || adapter.getSourceLevelURIs().contains(resource.getURI()))
+			return false;
+		return doesStorageExist(resource);
+	}
+
+	/**
+	 * Finds or creates a ResourceStorageLoadable for the given resource. Clients should first call
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageFacade.java:60]: `@link` on this and the next line please

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageLoadable.java [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+		} catch (ClassNotFoundException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	protected void readNodeModel(StorageAwareResource resource, InputStream inputStream) throws IOException {
+		SerializableNodeModel serializableNodeModel = new SerializableNodeModel(resource);
+		if (!resource.getResourceSet().getURIConverter().exists(resource.getURI(),
+				resource.getResourceSet().getLoadOptions())) {
+			ResourceStorageLoadable.LOG.info("Skipping loading node model for synthetic resource " + resource.getURI());
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageLoadable.java:105]: s/ResourceStorageLoadable.LOG/LOG/

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableEObjectDescription.java [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	@Override
+	public String getUserData(String key) {
+		return userData.get(key);
+	}
+
+	@Override
+	public String[] getUserDataKeys() {
+		return (String[]) Conversions.unwrapArray(userData.keySet(), String.class);
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableEObjectDescription.java:69]: You could use `userData.keySet().stream().toArray(String[]::new)`

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableReferenceDescription.java [hunk 27]
```diff
@@ -0,0 +1,99 @@
+/**
+ * Copyright (c) 2014, 2017 itemis AG (http://www.itemis.eu) and others.
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableReferenceDescription.java:2]: Update copyright

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableResourceDescription.java [hunk 28]
```diff
@@ -0,0 +1,165 @@
...
+			ref.updateResourceURI(uri, uRI);
+		for (SerializableEObjectDescription desc : descriptions)
+			desc.updateResourceURI(uri);
+		uRI = uri;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	protected List<IEObjectDescription> computeExportedObjects() {
+		return ((List<IEObjectDescription>) ((List<?>) this.descriptions));
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableResourceDescription.java:58]: I don't immediately say why the casts are needed? Isn't SerializableEObjectDescription instanceof IEObjectDescription

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableResourceDescription.java [hunk 29]
```diff
@@ -0,0 +1,165 @@
...
+
+	public void setURI(URI uRI) {
+		this.uRI = uRI;
+	}
+
+	public static SerializableResourceDescription createCopy(IResourceDescription desc) {
+		SerializableResourceDescription description = new SerializableResourceDescription();
+		description.setURI(desc.getURI());
+		description.descriptions = toList(
+				map(desc.getExportedObjects(), it -> SerializableResourceDescription.createCopy(it)));
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableResourceDescription.java:135]: SerializableResourceDescription::createCopy

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableResourceDescription.java [hunk 30]
```diff
@@ -0,0 +1,165 @@
...
+		description.references = toList(
+				map(desc.getReferenceDescriptions(), it -> SerializableResourceDescription.createCopy(it)));
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableResourceDescription.java:137]: SerializableResourceDescription::createCopy

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableResourceDescription.java (hunk 28 above)
Comment (szarnekow) [reply org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableResourceDescription.java:58]: True. But a `List< SerializableEObjectDescription>` is not a `List<IEObjectDescription>`.

**Comment by szarnekow** [comment]
@cdietrich it would be lame if I'd review this mayself now.

**Comment by tivervac** [comment]
Oh I see Sebastian actually force pushed instead of Arne, let me go over the old comments

**Comment by tivervac** [comment]
Seems like there are still 4 comments that went missing, see Sebastian's comments above

**Comment by szarnekow** [comment]
> Seems like there are still 4 comments that went missing, see Sebastian's comments above

Good catch, @tivervac - PR updated.

**Comment by cdietrich** [comment]
@szarnekow @tivervac what is the state of this pr?

**Comment by szarnekow** [comment]
From my perspective it's ready to merge. Given that I did the last round of commits here, I didn't want to approve it myself, though.

**Comment by tivervac** [comment]
I reviewed Sebastian's changes, LGTM @cdietrich.

@szarnekow You'll have to give your approve as you're still requesting changes

**Comment by cdietrich** [comment]
thx @ArneDeutsch @szarnekow

**Pull request merged**